modules/discretedata/connectivity/elementiterator.h

```diff
@@ -0,0 +1,143 @@
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*
+*********************************************************************************/
+
+#pragma once
+
+#include <discretedata/discretedatamoduledefine.h>
+#include <inviwo/core/common/inviwo.h>
+
+#include "connectivity.h"
+
+namespace inviwo {
+namespace dd {
+
+class ConnectionRange;
+
+/** \class ElementIterator
+*   Iterates over one GridPrimitive type in a Connectivity.
+*/
+class IVW_MODULE_DISCRETEDATA_API ElementIterator {
+    friend ElementIterator operator+(ind, ElementIterator&);
+    friend ElementIterator operator-(ind, ElementIterator&);
+
+public:
+    ElementIterator(const Connectivity* parent, GridPrimitive dimension, ind index = 0)
+        : index_(index), parent_(parent), dimension_(dimension) {}
+
+    ElementIterator() : index_(-1), parent_(nullptr), dimension_(GridPrimitive(-1)) {}
+
+    ~ElementIterator() = default;
+
+    /** Dereference to 'get data' */
+    ElementIterator operator*() const;
+
+    //*** Bidirectional Iteration ***\\
+
+    /** Walk forward */
+    ElementIterator& operator++() {
+        index_++;
+        return *this;
+    }
+
+    /** Walk backward */
+    ElementIterator& operator--() {
+        index_--;
+        return *this;
+    }
+
+    /** Compare */
+    bool operator==(ElementIterator& other) {
+        return other.parent_ == parent_  // Compare pointers.
+               && other.dimension_ == dimension_ && other.index_ == index_;
+    }
+
+    /** Compare */
+    bool operator!=(ElementIterator& other) { return !(other == *this); }
+
+    //*** Random Access Iteration ***\\
+
+    /** Increment randomly */
+    ElementIterator operator+(ind offset) {
+        return ElementIterator(parent_, dimension_, index_ + offset);
+    }
+
+    /** Increment randomly */
+    ElementIterator& operator+=(ind offset) {
+        index_ += offset;
+        return *this;
+    }
+
+    /** Decrement randomly */
+    ElementIterator operator-(ind offset) {
+        return ElementIterator(parent_, dimension_, index_ - offset);
+    }
+
+    /** Decrement randomly */
+    ElementIterator& operator-=(ind offset) {
+        index_ -= offset;
+        return *this;
+    }
+
+    /** GridPrimitive type the iterator walks through */
+    GridPrimitive getType() const { return dimension_; }
+
+    /** The current index. Equivalent to dereferencing. */
+    ind getIndex() const { return index_; }
+
+    /** Iterate over connected GridPrimitives (neighbors etc) */
+    ConnectionRange connection(GridPrimitive toType) const;
+
+    // Members
+protected:
+    /** Index to the current element */
+    ind index_;
+
+    /** Pointer to Connectivity iterated through - Does not delete */
+    const Connectivity* parent_;
+
+    /** GridPrimitive type iterated over (0D vertices etc) */
+    const GridPrimitive dimension_;
```

Comment:
Question it the GridPrimitive the same as the Polytope (https://en.wikipedia.org/wiki/Polytope)? I think it would be nice if we could find a different naming scheme here then using dimension, since it's meaning is a bit ambiguous

Reply:
or is a Simplex type? https://en.wikipedia.org/wiki/Simplex there are so many names for these things :)